Align tags to a column, and support org-indent-mode's virtual indentation

Sodel-the-Vociferous opened the pull request:
Orgzly and Emacs re-aligning the tags all the time made keeping my notes in source control bothersome. (I had to resolve lots of conflicts.) 

Now, setting OrgParserSettings.tagsColumn to -77 matches org-mode's default behaviour. Setting it to 0 matches the previous behavior.

I changed the default to mimic Emacs (tagsColumn=-77), but I don't mind changing it back to 0 if you prefer, so the default behavior doesn't change.

if this PR gets accepted, I'll look at exposing this setting in Orgzly's settings menu too.

nevenz commented:
I think default should definitely be kept the same (0), so we don't break anything for existing Orgzly users.

I haven't checked the code yet, but one potential issue that comes to mind is using `indent` option in Org mode. Tags will be shifted left if it is used to account for its "virtual indent". There are options in Org mode that also affect the amount (like indentation per level), which can make this issue kinda tricky.

I remember wishing (when I consider supporting this one) that tags indentation is also virtual in Org mode, if possible (and fast enough).

Anyway, do you think this can be easily handled? We might want to start from listing all the possible scenarios.

Thanks for working on this.

Sodel-the-Vociferous commented:
Good point. I have changed tagsColumn's default to 0.

Well! I didn't know org-indent-mode even existed! Thanks for bringing it up, I think it will be useful to me.

I looked at the code for org-indent-mode, and I think it can be handled without much trouble. The virtual indentation that org-indent-mode adds is a simple offset. I added `OrgParserSettings.realPlusVirtualIndentationPerLevel`, which should act the same as `org-indent-indentation-per-level`.

Since there is already one level of indentation (the heading's stars), I shift the tags to the left by:
`(level * (indentation per level - 1) )`

And, that's all org-indent-mode does. :)

nevenz commented:
I'm not getting the same results as in Org mode. Tried with `org-tags-column 20` and `org-indent-indentation-per-level 4`, I get...

In Org mode:

```
* A                 :a:                                                                             
** B             :b:                                                                             
*** C         :c:  
```

Export from Orgzly:

```
* A              :a:
** B          :b:
*** C      :c:
```

We probably want few tests written for different values (positive and negative `org-tags-column`, `org-indent-indentation-per-level` `0` or positive, multiple tags with different length). Check few tests in `OrgParserTest` that are using `OrgParserSettings`.

BTW, `org-indent-indentation-per-level` is `2` by default.

Sodel-the-Vociferous commented:
You're right, sorry about that. I forgot that org-indent-mode doesn't add any offset to top-level headings.
So, the left-offset to the tags when virtual indentation is used is:
(level - 1) * (indentation per level - 1)

And, I added a few unit tests, too, to make sure Orgzly's output will still match org-mode's.

nevenz commented:
Great, thanks!

I've just cleaned a code a bit to reuse the new methods you've created (`toString` and `parsedWithSettings`).
